pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -197,179 +211,108 @@ func (r *repository) Get(dgst digest.Digest) (*schema1.SignedManifest, error) {
 		return nil, err
 	}
 
+	supportsSchema2 := r.isSchema2Supported(ctx)
+
 	ref := imageapi.DockerImageReference{Namespace: r.namespace, Name: r.name, Registry: r.registryAddr}
-	return r.manifestFromImageWithCachedLayers(image, ref.DockerClientDefaults().Exact())
-}
+	manifest, err := r.manifestFromImageWithCachedLayers(image, ref.DockerClientDefaults().Exact(), supportsSchema2)
 
-// Enumerate retrieves digests of manifest revisions in particular repository
-func (r *repository) Enumerate() ([]digest.Digest, error) {
-	panic("not implemented")
+	return manifest, err
 }
 
-// GetByTag retrieves the named manifest with the provided tag
-func (r *repository) GetByTag(tag string, options ...distribution.ManifestServiceOption) (*schema1.SignedManifest, error) {
-	for _, opt := range options {
-		if err := opt(r); err != nil {
-			return nil, err
-		}
-	}
-
-	// find the image mapped to this tag
-	imageStreamTag, err := r.getImageStreamTag(tag)
-	if err != nil {
-		// TODO: typed errors
-		context.GetLogger(r.ctx).Errorf("Error getting ImageStreamTag %q: %v", tag, err)
-		return nil, err
-	}
-	image := &imageStreamTag.Image
-
-	ref, referenceErr := imageapi.ParseDockerImageReference(image.DockerImageReference)
-	if referenceErr == nil {
-		ref.Namespace = r.namespace
-		ref.Name = r.name
-		ref.Registry = r.registryAddr
-	}
-	defaultRef := ref.DockerClientDefaults()
-	cacheName := defaultRef.AsRepository().Exact()
-
-	// if we have a local manifest, use it
-	if len(image.DockerImageManifest) > 0 {
-		return r.manifestFromImageWithCachedLayers(image, cacheName)
-	}
-
-	dgst, err := digest.ParseDigest(imageStreamTag.Image.Name)
-	if err != nil {
-		context.GetLogger(r.ctx).Errorf("Error parsing digest %q: %v", imageStreamTag.Image.Name, err)
-		return nil, err
-	}
-
-	if localImage, err := r.getImage(dgst); err != nil {
-		// if the image is managed by OpenShift and we cannot load the image, report an error
-		if image.Annotations[imageapi.ManagedByOpenShiftAnnotation] == "true" {
-			context.GetLogger(r.ctx).Errorf("Error getting image %q: %v", dgst.String(), err)
-			return nil, err
+func (r *repository) isSchema2Supported(ctx context.Context) bool {
+	if req, err := context.GetRequest(ctx); err == nil {
+		if acceptHeaders, ok := req.Header["Accept"]; ok {
+			for _, mediaType := range acceptHeaders {
+				if mediaType == schema2.MediaTypeManifest {
+					return true
```

Comment:
Too many levels. Please rewrite to return-early style. E.g.

    req, err := context.GetRequest(ctx)
    if err != nil {
        return false
    }